### PR TITLE
Show mod path on hover

### LINKED ISSUE #1064: Hover should display module path for item
Sometimes when hovering over an item I wonder where it came from. It would be nice to display the path to the item somewhere (maybe above or below the decl).

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Closes #1064

### CODE REVIEW

File: crates/ra_project_model/src/lib.rs [hunk 1]
```diff
@@ -191,6 +190,8 @@ impl ProjectWorkspace {
                             crate_graph.add_crate_root(
                                 file_id,
                                 edition,
+                                // FIXME json definitions can store the crate name
```
Comment (edwin0cheng) [inline crates/ra_project_model/src/lib.rs]: Yeah, I am working on vscode config for `OUT_DIR` and just find out there is no crate name in json project definitions. :)

File: crates/ra_db/src/input.rs [hunk 2]
```diff
@@ -107,6 +107,7 @@ impl CrateName {
 struct CrateData {
     file_id: FileId,
     edition: Edition,
+    declaration_name: Option<String>,
```
Comment (matklad) [inline crates/ra_db/src/input.rs]:
```suggestion
    display_name: Option<String>,
```

Comment (matklad) [reply crates/ra_db/src/input.rs]: IE, the only purpose of this name is to be displayed to the user.

File: crates/ra_ide/src/hover.rs [hunk 3]
```diff
@@ -83,44 +89,86 @@ impl HoverResult {
...
+        once(db.crate_graph().declaration_name(&module.krate().into()).cloned())
+            .chain(
+                module
+                    .path_to_root(db)
+                    .into_iter()
+                    .rev()
+                    .map(|it| it.name(db).map(|name| name.to_string())),
+            )
+            .chain(once(definition_owner_name(db, def)))
+            .filter_map(std::convert::identity)
```
Comment (matklad) [inline crates/ra_ide/src/hover.rs]: https://doc.rust-lang.org/stable/core/iter/trait.Iterator.html#method.flatten

File: crates/ra_db/src/input.rs [hunk 4]
```diff
@@ -177,6 +179,15 @@ impl CrateGraph {
         self.arena[&crate_id].edition
     }
 
+    /// Returns a name of a crate, declared in the root project.
+    /// May be missing for some cases, such as when the crate definition was created for a code snippet.
+    ///
+    /// This should not be considered as a normal crate name, since the actual name can be different in
+    /// a particular dependent crate, where it is specified.
+    pub fn declaration_name(&self, crate_id: &CrateId) -> Option<&String> {
```
Comment (matklad) [inline crates/ra_db/src/input.rs]:
```suggestion
    pub fn declaration_name(&self, crate_id: &CrateId) -> Option<&str> {
```

`&String` is one of the "useless" types in Rust, along with `&Vec<T>`, `&Box<T>`, etc. 

Also, I feel we should redo this API to try to be less abstract, and just expose CrateData:

```
crate_graph[crate_id].deps()
```

But this is unrelated to the PR.

Comment (SomeoneToIgnore) [reply crates/ra_db/src/input.rs]: >  we should redo this API to try to be less abstract, and just expose CrateData

That was easy to do, so why not.

File: crates/ra_db/src/input.rs (hunk 2 above)
Comment (SomeoneToIgnore) [reply crates/ra_db/src/input.rs]: Nice name for a name, thanks.
